cli/args/lockfile.rs

```diff
@@ -127,23 +130,52 @@ impl CliLockfile {
     };
 
     let lockfile = if flags.lock_write {
-      CliLockfile::new(Lockfile::new_empty(filename, true))
+      CliLockfile::new(
+        Lockfile::new_empty(filename, true),
+        flags.frozen_lockfile,
+      )
     } else {
-      Self::read_from_path(filename)?
+      Self::read_from_path(filename, flags.frozen_lockfile)?
     };
     Ok(Some(lockfile))
   }
-  pub fn read_from_path(filename: PathBuf) -> Result<CliLockfile, AnyError> {
+  pub fn read_from_path(
+    filename: PathBuf,
+    frozen: bool,
+  ) -> Result<CliLockfile, AnyError> {
     match std::fs::read_to_string(&filename) {
-      Ok(text) => Ok(CliLockfile::new(Lockfile::with_lockfile_content(
-        filename, &text, false,
-      )?)),
-      Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
-        Ok(CliLockfile::new(Lockfile::new_empty(filename, false)))
-      }
+      Ok(text) => Ok(CliLockfile::new(
+        Lockfile::with_lockfile_content(filename, &text, false)?,
+        frozen,
+      )),
+      Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(
+        CliLockfile::new(Lockfile::new_empty(filename, false), frozen),
+      ),
       Err(err) => Err(err).with_context(|| {
         format!("Failed reading lockfile '{}'", filename.display())
       }),
     }
   }
+  pub fn error_if_changed(&self) -> Result<(), AnyError> {
+    let lockfile = self.lockfile.lock();
+    if self.frozen && lockfile.has_content_changed {
+      let suggested = if *super::DENO_FUTURE {
+        "`deno cache`, `deno install`,"
+      } else {
+        "`deno cache`"
+      };
+
+      let contents =
+        std::fs::read_to_string(&lockfile.filename).unwrap_or_default();
+      let new_contents = lockfile.as_json_string();
+      let diff = crate::util::diff::diff(&contents, &new_contents);
+      // has an extra newline at the end
+      let diff = diff.trim_end();
+      Err(deno_core::anyhow::anyhow!(
+        "The lockfile is out of date. Run {suggested} or rerun without the `--frozen` flag to update it.\nchanges:\n{diff}"
```

Comment:
Should mention `--frozen=false` as well?

Reply:
Updated